keps/sig-scheduling/3022-tuning-the-number-of-domains-on-pod-topology-spread/README.md

```diff
@@ -0,0 +1,510 @@
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests for meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+With [Pod Topology Spread](/keps/sig-scheduling/895-pod-topology-spread), users can define the rule to spread pods across your cluster among failure-domains.
+And, we propose to add the parameters `minDomains` and `maxDomains` to control the number of domains on Pod Topology Spread.
+
+## Motivation
+
+Pod Topology Spread has [`maxSkew` parameter](https://github.com/kubernetes/enhancements/tree/11a976c74e1358efccf251d4c7611d05ce27feb3/keps/sig-scheduling/895-pod-topology-spread#maxskew), which control the degree to which Pods may be unevenly distributed.
+But, there isn't a way to control the number of domains over which we should spread.
+
+### Goals
+
+Users can define a maximum number of domains with `maxDomains` parameter
+
+When the number of Pods is more than `maxDomains`,
+
+- when `whenUnsatisfiable` equals `DoNotSchedule`, scheduler schedules each Pods to a different Node.
+  i.e. scheduler filters Nodes that don't have matching Pods in the `Filter` phase.
+- when `whenUnsatisfiable` equals `ScheduleAnyway`, scheduler prefers Nodes that have matching Pods.
+  i.e. scheduler gives a high score to Nodes that have matching Pods in the `Score` phase.
+
+And, Users can define a minimum number of domains with `minDomains` parameter
+
+When the number of Pods is less than `minDomains`,
+
+- when `whenUnsatisfiable` equals `DoNotSchedule`, scheduler doesn't schedule a matching Pod to Nodes that have matching Pods.
+  i.e. scheduler filters Nodes that have matching Pods in the `Filter` phase.
+- when `whenUnsatisfiable` equals `ScheduleAnyway`, scheduler prefers Nodes that don't have matching Pods.
```

Comment:
### Alternatives

When users use `minDomains` and `whenUnsatisfiable=DoNotSchedule`, and the number of Pods is less than `minDomains`, scheduler waits to schedule Pods to each Nodes.
For example, when a user sets `minDomains=5` and `whenUnsatisfiable=DoNotSchedule`, then creates only 4 Pods, all Pods won't be scheduled until the fifth Pod is created.
When the fifth pod is created, bind all pods.

---

Which do you think is better?